logstash-core/src/main/java/org/logstash/common/io/ByteBufferPageIO.java

```diff
@@ -0,0 +1,248 @@
+        return persistedByteCount(bytes) <= bytesLeft;
+    }
+
+    @Override
+    public void write(byte[] bytes, long seqNum) throws IOException {
+        // since writes always happen at head, we can just append head to the offsetMap
+        assert this.offsetMap.size() == this.elementCount :
+                String.format("offsetMap size=%d != elementCount=%d", this.offsetMap.size(), this.elementCount);
+
+        int initialHead = this.head;
+
+        this.buffer.position(this.head);
+        this.buffer.putLong(seqNum);
+        this.buffer.putInt(bytes.length);
+        this.buffer.put(bytes);
+        this.buffer.putInt(checksum(bytes));
+        this.head += persistedByteCount(bytes.length);
+        assert this.head == this.buffer.position() :
+                String.format("head=%d != buffer position=%d", this.head, this.buffer.position());
+
+        if (this.elementCount <= 0) {
+            this.minSeqNum = seqNum;
+        }
+        this.offsetMap.add(initialHead);
+        this.elementCount++;
+    }
+
+    @Override
+    public SequencedList<byte[]> read(long seqNum, int limit) throws IOException {
+        assert seqNum >= this.minSeqNum :
+                String.format("seqNum=%d < minSeqNum=%d", seqNum, this.minSeqNum);
+        assert seqNum <= maxSeqNum() :
+                String.format("seqNum=%d is > maxSeqNum=%d", seqNum, maxSeqNum());
+
+        List<byte[]> elements = new ArrayList<>();
+        List<Long> seqNums = new ArrayList<>();
+
+        int offset = this.offsetMap.get((int)(seqNum - this.minSeqNum));
+
+        this.buffer.position(offset);
+
+        for (int i = 0; i < limit; i++) {
+            long readSeqNum = this.buffer.getLong();
+
+            assert readSeqNum == (seqNum + i) :
+                    String.format("unmatched seqNum=%d to readSeqNum=%d", seqNum + i, readSeqNum);
+
+            int readLength = this.buffer.getInt();
+            byte[] readBytes = new byte[readLength];
+            this.buffer.get(readBytes);
+            int checksum = this.buffer.getInt();
+            int computedChecksum = checksum(readBytes);
+            if (computedChecksum != checksum) {
+                throw new IOException(String.format("computed checksum=%d != checksum for file=%d", computedChecksum, checksum));
+            }
+
+            elements.add(readBytes);
+            seqNums.add(readSeqNum);
+
+            if (seqNum + i >= maxSeqNum()) {
+                break;
+            }
+        }
+
+        return new SequencedList<>(elements, seqNums);
+    }
+
+    @Override
+    public void deactivate() {
+        // nothing to do
+    }
+
+    @Override
+    public void activate() {
+        // nothing to do
+    }
+
+    @Override
+    public void ensurePersisted() {
+        // nothing to do
+    }
+
+    @Override
+    public void purge() throws IOException {
+        // do nothing
+    }
+
+    @Override
+    public void close() throws IOException {
+        // TBD
```

Comment:
this means it should be implemented at some point if needed. not sure we need to do anything since this is the in-memory page implementation which is only used for tests for now.